### PR TITLE
Add Pub/Sub Java client API example.

### PR DESCRIPTION
**Pull request by sabhyankar**

### CODE REVIEW

File: examples/pubsub-publish-avro-example/README.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+* Cloud Pub/Sub topic and subscription
+    * The Pub/Sub topic will be used by the client to publish messages.
+    * The Pub/Sub subscription on the same topic will be used by the Dataflow job to read the messages.
+* BigQuery table to stream records into - The table schema should match the Avro schema of the messages.
+
+### Building the Project
+
+Build the entire project using the maven compile command.
+```sh
+mvn clean && mvn compile
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/README.md]: Change to 

```sh
mvn clean compile
```

File: examples/pubsub-publish-avro-example/README.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+### Running unit tests
+Run the unit tests using the maven test command.
+```sh
+mvn clean && mvn test
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/README.md]: Change to 

```sh
mvn clean test
```

File: examples/pubsub-publish-avro-example/README.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+### Publishing sample records to Cloud Pub/Sub
+Publish sample [Employee](src/main/avro/employee.avsc) records using the maven exec command.
+```sh
+mvn compile exec:java \
+-Dexec.mainClass=com.google.cloud.pso.pubsub.EmployeePublisherMain \
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/README.md]: Add 2 spaces on parameter lines for readability

File: examples/pubsub-publish-avro-example/README.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      Default: PT0.5S
+    --elementCount, -c
+      The number of elements to be batched in each request.
+      Default: 500
+    --help, -h
+
+    --numberOfMessages, -n
+      Number of sample messages to publish to Pub/Sub
+      Default: 100000
+  * --topic, -t
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/README.md]: Reason for star?

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pipeline/IngestionMain.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    Pipeline pipeline = Pipeline.create(options);
+
+    /*
+     * Steps:
+     *  1) Read from Pub/Sub
+     *  2) Stream into BigQuery
+     */
+
+    pipeline
+
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pipeline/IngestionMain.java]: Remove new-line

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pipeline/IngestionMain.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    // Create the pipeline
+    Pipeline pipeline = Pipeline.create(options);
+
+    /*
+     * Steps:
+     *  1) Read from Pub/Sub
+     *  2) Stream into BigQuery
+     */
+
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pipeline/IngestionMain.java]: Remove new-line

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/EmployeePublisher.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+/**
+ * An implementation of the {@link ObjectPublisher} to publish test {@link Employee} objects to
+ * Pub/Sub
+ */
+public class EmployeePublisher extends ObjectPublisher<Employee> {
+
+  // Since the primary consumer of these records would be a Dataflow pipeline that will
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/EmployeePublisher.java]: Use mult-line Java comment 

```java
/*
 *
 */
```

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/EmployeePublisherMain.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    try {
+      jCommander.parse(args);
+    } catch (ParameterException e) {
+      jCommander.usage();
+      return;
+    }
+
+    if (parsedArgs.isHelp()) {
+      jCommander.usage();
+      return;
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/EmployeePublisherMain.java]: Remove return and place the following block in an `else` statement

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/EmployeePublisherMain.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    // For demonstration purposes we will use an implementation of ObjectReader
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/EmployeePublisherMain.java]: Use multi-line Java comment 

```java
/*
 *
 */
```

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/EmployeePublisherMain.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    // to generate random Employee objects to be published.
+    ObjectReader<Employee> employeeReader =
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/EmployeePublisherMain.java]: Maybe encapsulate the following into a private `publish()` method taking in the `parsedArgs` as a parameter

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A generic class that can be used to publish any object to Pub/Sub with configurable {@link
+ * BatchingSettings}
+ *
+ * @param <T>
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java]:
```java
@param <T> -  The type of objects to publish
```

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+ */
+public abstract class ObjectPublisher<T> {
+  private static final long DEFAULT_THREAD_SLEEP = 1000L;
+  private com.google.cloud.pubsub.v1.Publisher publisher;
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java]: Import?

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+  private AtomicLong awaitedFutures;
+  private ExecutorService executor = Executors.newCachedThreadPool();
+
+  private void init(Args args) {
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java]: Add Javadoc

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+            .setBatchingSettings(batchingSettings);
+    try {
+      this.publisher = builder.build();
+    } catch (IOException e) {
+      System.out.println("Could not create publisher: " + e);
+      System.exit(-1);
+    }
+  }
+
+  private void publish(T record) throws IOException {
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java]: Add Javadoc

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+  /**
+   * Method to serialize the object to be published into a ByteString.
+   *
+   * @param record object to serliaze into a ByteString
+   * @return ByteString representation of the input record
+   * @throws IOException
+   */
+  public abstract ByteString serialize(T record) throws IOException;
+
+  private void waitForPublishes() {
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java]: Add Javadoc.

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+    try {
+      while (awaitedFutures.longValue() > 0) {
+        Thread.sleep(DEFAULT_THREAD_SLEEP);
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java]: Use

```java
TimeUnit.MILLISECONDS.sleep(..)
```

File: examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+    public Long getNumOfMessages() {
+      return numOfMessages;
+    }
+
+    public boolean isHelp() {
+      return help;
+    }
+  }
+
+  private static class DurationConverter implements IStringConverter<Duration> {
```
Comment (ryanmcdowell) [inline examples/pubsub-publish-avro-example/src/main/java/com/google/cloud/pso/pubsub/common/ObjectPublisher.java]: Add Javadoc.

File: examples/pubsub-publish-avro-example/README.md (hunk 4 above)
Comment (sabhyankar) [reply examples/pubsub-publish-avro-example/README.md]: This is an artifact of JCommander's default usage display. * indicates a required parameter.

**Pull request merged**